ouroboros-consensus-test/PLAN-statistical-testing.md

```diff
@@ -0,0 +1,159 @@
+    violation, if any. Include a label when that happens. Occasionally review
+    those label rates, comparing them to the baseline simple small test
+    configurations.
+
+  - These non-statistical properties can use values of k, f, etc that are less
+    "small" so that we truncate less often. But they'll still be "large" enough
+    that the CP violation rate is non-trivial. (EG I'm hoping for something
+    around 1-5%.)
+
+That's the primary idea. It's focused on CP violations, but I anticipate we'll
+be able to re-use the same iterations for testing multiple properties, such as
+whether f actually induces a correct fraction of active slots.
+
+## Steps
+
+  - I'm starting with Mock Praos, because it is simpler and especially because
+    it has no stability window limitation. Once those tests are proved out and
+    relatively mature, I'll replicate it for Shelley, and then eventually
+    Cardano.
+
+  - Detect CP violations so that we can count them, label them, truncate the
+    test, etc. Fail fatally if the leader schedule should have precluded CP
+    violations (eg there were no multi-leader slots).
+
+  - Empirically determine the CP violation. Use eg 100,000 iterations. (I hope
+    QuickCheck's tabulate can withstand that...)
+
+  - Implement a dedicated statistical property test. (I anticipate
+    `checkCoverage` and `coverage` and sufficiently well-behaved when used in
+    isolation.)
+
+  - Tune the confidence-level and also potentially invest time to optimize the
+    relevant mock/test infrastructure (at least check for low-hanging fruit).
+
+  - Label CPs and truncate for CPs in the Mock Praos non-statistical
+    properties. (Maybe also Shelley, even before we have dedicated Shelley
+    statistical tests?)
+
+  - Introduce semi-synchrony, ie delta greater than 1. I'm disappointed that I
+    still don't have confidence regarding how exactly to inject these delays.
+    My only insight so far that feels particularly helpful is to only delay
+    RollForward messages and let all others traveral "instantly" (this is
+    already implemented and used by the Cardano network partition). RollForward
+    is the mini-protocol that best corresponds to the "entire chain as a
+    message" used in the Ouroboros papers. However, many questions remain. EG
+    our reliable order-preserving channels can cause messages to get stuck
+    behind other messages, and -- similarly to priority inversion -- this may
+    delay a particular message beyond it's originally scheduled delay.
+
+## Miscellaneous Questions
+
+  - The above proposes focusing on the probability of a CP violation occuring
+    during some fixed number of slots. Should we instead focus on its arrival
+    rate? This would suggest choosing a "small" configuration that essentially
+    ensures a CP violation will occur within say the first 1000 slots --
+    otherwise the test may have extremely variable run-times.
```

Comment:
I'm not sure what the consequences, can you explan?

Reply:
I added a sentence here that explains that I'm currently not sure how to elaborate :)